src/Storages/S3Queue/S3QueueFilesMetadata.cpp

```diff
@@ -28,326 +42,875 @@ namespace
 {
-    auto zookeeper = storage->getZooKeeper();
-    String processing_files;
-    zookeeper->tryGet(zookeeper_processing_path, processing_files);
+    auto timer = DB::CurrentThread::getProfileEvents().timer(ProfileEvents::S3QueueSetFileFailedMicroseconds);
+    const auto & path = holder->path;
+
+    auto file_status = local_file_statuses.get(path, /* create */false);
+    {
+        std::lock_guard lock(file_status->metadata_lock);
+        file_status->state = FileStatus::State::Failed;
+        file_status->last_exception = exception_message;
+        file_status->processing_end_time = std::chrono::system_clock::to_time_t(std::chrono::system_clock::now());
+    }
+
+    SCOPE_EXIT({
+        file_status->profile_counters.increment(ProfileEvents::S3QueueSetFileFailedMicroseconds, timer.get());
+        timer.cancel();
+    });
 
-    S3QueueProcessingCollection processing_collection;
-    if (!processing_files.empty())
-        processing_collection.parse(processing_files);
-    processing_collection.add(file_paths);
+    const auto node_name = getNodeName(path);
+    auto node_metadata = createNodeMetadata(path, exception_message);
+    const auto zk_client = getZooKeeper();
 
-    if (zookeeper->exists(zookeeper_processing_path))
-        zookeeper->set(zookeeper_processing_path, processing_collection.toString());
+    /// Is file retriable?
+    if (max_loading_retries == 0)
+    {
+        /// File is not retriable,
+        /// just create a node in /failed and remove a node from /processing.
+
+        Coordination::Requests requests;
+        requests.push_back(zkutil::makeCreateRequest(zookeeper_failed_path / node_name,
+                                                     node_metadata.toString(),
+                                                     zkutil::CreateMode::Persistent));
+        Coordination::Responses responses;
+        if (holder->remove(&requests, &responses))
+        {
+            LOG_TEST(log, "File `{}` failed to process and will not be retried. "
+                     "Error: {}", path, exception_message);
+            return;
+        }
+
+        if (responses[0]->error != Coordination::Error::ZOK)
+        {
+            throw Exception(ErrorCodes::LOGICAL_ERROR,
+                            "Cannot create a persistent node in /failed since it already exists");
+        }
+
+        LOG_WARNING(log, "Cannot set file ({}) as processed since processing node "
+                    "does not exist with expected processing id does not exist, "
+                    "this could be a result of expired zookeeper session", path);
+        return;
+    }
+
+    /// So file is retriable.
+    /// Let's do an optimization here.
+    /// Instead of creating a persistent /failed/node_hash node
+    /// we create a persistent /failed/node_hash.retriable node.
+    /// This allows us to make less zookeeper requests as we avoid checking
+    /// the number of already done retries in trySetFileAsProcessing.
```

Comment:
AFAIU, In case when we succeded after some retry, we don't delete nodes `/failed/node_hash.retriable`, so they can accumulate there and comsume some memory in keeper.